src/NServiceBus.Transport.Msmq/DelayedDelivery/DueDelayedMessagePoller.cs

```diff
@@ -0,0 +1,341 @@
+        }
+
+        public async Task Stop(CancellationToken cancellationToken = default)
+        {
+            tokenSource.Cancel();
+            await loopTask.ConfigureAwait(false);
+            await completionTask.ConfigureAwait(false);
+        }
+
+        public void Signal(DateTimeOffset timeoutTime)
+        {
+            //If the next timeout is within a minute from now, trigger the poller
+            if (DateTimeOffset.UtcNow.Add(MaxSleepDuration) > timeoutTime)
+            {
+                //If there is something already in the queue we are fine.
+                signalQueue.Writer.TryWrite(true);
+            }
+        }
+
+        /// <summary>
+        /// This method does not accept a cancellation token because it need to finish all the tasks that have been started even if cancellation is under way
+        /// </summary>
+#pragma warning disable PS0018 // A task-returning method should have a CancellationToken parameter unless it has a parameter implementing ICancellableContext
+        async Task AwaitHandleTasks()
+#pragma warning restore PS0018 // A task-returning method should have a CancellationToken parameter unless it has a parameter implementing ICancellableContext
+        {
+            while (await taskQueue.Reader.WaitToReadAsync().ConfigureAwait(false)) //if this returns false the channel is completed
+            {
+                while (taskQueue.Reader.TryRead(out var task))
+                {
+                    try
+                    {
+                        await task.ConfigureAwait(false);
+                    }
+                    catch (Exception e)
+                    {
+                        failureHandlingCircuitBreaker.Failure(e);
+                    }
+                }
+            }
+        }
+
+        async Task Loop(CancellationToken cancellationToken)
+        {
+            try
+            {
+                while (!cancellationToken.IsCancellationRequested)
+                {
+#pragma warning disable PS0021 // Highlight when a try block passes multiple cancellation tokens
+                    try
+#pragma warning restore PS0021 // Highlight when a try block passes multiple cancellation tokens
+                    {
+                        var completionSource = new TaskCompletionSource<DateTimeOffset?>();
+                        var handleTask = Poll(completionSource, cancellationToken);
+                        await taskQueue.Writer.WriteAsync(handleTask, cancellationToken).ConfigureAwait(false);
+                        var nextPoll = await completionSource.Task.ConfigureAwait(false);
+
+                        if (nextPoll.HasValue)
+                        {
+                            using (var waitCancelled = new CancellationTokenSource())
+                            {
+                                var combinedSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken, waitCancelled.Token, cancellationToken);
```

Comment:
Seems like a typo? As `cancellationToken` is passed twice?

Reply:
No, it's not, this is creating a combined source based on multiple tokens so that if one of them is canceled, the source is canceled too.

Reply:
Actually it is a mistake. I added it twice for some reason. I was probably really tired.